src/sdam/monitor.rs

```diff
@@ -56,84 +86,109 @@ pub(super) fn monitor_server(
+        };
+
+        // Send an isMaster to the server.
+        let server_description = self.check_server();
+        self.server_type = server_description.server_type;
+
+        update_topology(topology, server_description);
+    }
 
-fn check_server(
-    conn: &mut Connection,
-    server_type: ServerType,
-    server: &Arc<Server>,
-) -> ServerDescription {
-    let address = conn.address().clone();
+    fn check_server(&mut self) -> ServerDescription {
+        let address = self.address.clone();
 
-    match is_master(conn) {
-        Ok(reply) => return ServerDescription::new(address, Some(Ok(reply))),
-        Err(e) => {
-            server.clear_connection_pool();
+        match self.is_master() {
+            Ok(reply) => return ServerDescription::new(address, Some(Ok(reply))),
+            Err(e) => {
+                self.clear_connection_pool();
 
-            if server_type == ServerType::Unknown {
-                return ServerDescription::new(address, Some(Err(e)));
+                if self.server_type == ServerType::Unknown {
+                    return ServerDescription::new(address, Some(Err(e)));
+                }
             }
         }
+
+        ServerDescription::new(address, Some(self.is_master()))
     }
 
-    ServerDescription::new(address, Some(is_master(conn)))
-}
+    fn is_master(&mut self) -> Result<IsMasterReply> {
+        let result = self.is_master_inner();
+
+        if result
+            .as_ref()
+            .err()
+            .map(|e| e.kind.is_network_error())
+            .unwrap_or(false)
+        {
+            self.connection = self.make_new_connection()?;
+        }
+
+        result
+    }
+
+    fn is_master_inner(&mut self) -> Result<IsMasterReply> {
+        let command = Command::new_read(
+            "isMaster".into(),
+            "admin".into(),
+            None,
+            doc! { "isMaster": 1 },
+        );
+
+        let start_time = PreciseTime::now();
+        let command_response = self.connection.send_command(command, None)?;
+        let end_time = PreciseTime::now();
+
+        let command_response = command_response.body()?;
 
-fn is_master(conn: &mut Connection) -> Result<IsMasterReply> {
-    let command = Command::new_read(
-        "isMaster".into(),
-        "admin".into(),
-        None,
-        doc! { "isMaster": 1 },
-    );
-
-    let start_time = PreciseTime::now();
-    let command_response = conn.send_command(command, None)?;
-    let end_time = PreciseTime::now();
-
-    let command_response = command_response.body()?;
-
-    Ok(IsMasterReply {
-        command_response,
-        // TODO RUST-193: Round-trip time
-        round_trip_time: Some(start_time.to(end_time).to_std().unwrap()),
-    })
+        Ok(IsMasterReply {
+            command_response,
+            round_trip_time: Some(start_time.to(end_time).to_std().unwrap()),
+        })
+    }
+
+    fn make_new_connection(&self) -> Result<Connection> {
+        Connection::new(
+            0,
```

Comment:
The id and generation being hard coded to 0 seems like an issue. We should probably get these from the pool.

Reply:
Per the SDAM and CMAP specs, [monitoring connections cannot be from the regular connection pool on the server](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#a-monitor-should-not-use-the-client-s-regular-connection-pool), and [connection pools must start their auto-incrementing ID at 1](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#connection). Given that no CMAP events are emitted from the SDAM monitoring connections, the connection ID doesn't really matter, but if it did, picking 0 guarantees that it will never overlap with any of the IDs from the pool for the same server (unless we have more than `usize::MAX` connections, but that would cause overlap regardless of how we picked the ID here).

Reply:
Good idea; done